core/trino-main/src/main/java/io/trino/metadata/InMemoryNodeManager.java

```diff
@@ -73,7 +73,7 @@ public Set<InternalNode> getNodes(NodeState state)
     @Override
     public Set<InternalNode> getActiveCatalogNodes(CatalogHandle catalogHandle)
     {
-        return allNodes;
+        return ImmutableSet.copyOf(allNodes);
```

Comment:
same in `public Set<InternalNode> getNodes(NodeState state)`

also same when constructing `return new NodesSnapshot(` / `return new AllNodes(` -- these are separate classes defined in a different file, we shouldn't assume they do defensive copy in their constructors.